### PR TITLE
Customize arguments-differ error messages

### PR DESCRIPTION
**Pull request by ksaketou**
## Steps

- [ ] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description
This pull request customizes the output messages of `arguments-differ` error message based on different error cases (type change, name change, number of parameters etc). It aims to act as a base for the creation of a new error message, `arguments-renamed`.
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Related Issue

Part one of #3536

### CODE REVIEW

File: pylint/checkers/classes.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes.py:275]: Here it was a boolean before it look like it's now a list of string.
```suggestion
) -> List[str]:
```

File: pylint/checkers/classes.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes.py:309]:
```suggestion
                f"Parameter '{original_param.name}' has been renamed to '{overridden_param.name}' in"
```

File: pylint/checkers/classes.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes.py:319]:
```suggestion
) -> List[str]:
```

File: pylint/checkers/classes.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes.py:368]:
```suggestion
    if different_kwonly:
```

File: tests/functional/a/arguments_differ.py
Comment (Pierre-Sassoulas) [inline tests/functional/a/arguments_differ.py:6]: Could we refrain from changing the existing functional test in this MR, please ? I would like to see if the old use cases are all handled well. We can add more functional tests if required :)

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you, great work ! I suggested some small changes, the main thing being we should not change the existing tests, only their outputs. They're testing some use case that were constructed from dozen of issues and we need them to not regress. We can add other tests if required for typing check, having long functional test files with duplication is not a problem at all :)

**Comment by ksaketou** [comment]
Thank you! Okay I see, I'm working on these changes :)

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you , great work and very nice new message ! Could we also keep the typing change in functional tests for the next MR, please?

**Comment by ksaketou** [comment]
> Thank you , great work and very nice new message ! Could we also keep the typing change in functional tests for the next MR, please?

Thank you! What do you mean exactly? The parameter types on the tests?

File: tests/functional/a/arguments_differ.py
Comment (Pierre-Sassoulas) [inline tests/functional/a/arguments_differ.py:30]: Sorry for not being very clear ;) For example here we should keep this change (adding typing) for step 2.
```suggestion
    def func(cls, data):
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: tests/functional/a/arguments_differ.py
Comment (Pierre-Sassoulas) [reply tests/functional/a/arguments_differ.py:30]: (No change for python files in ``tests/functional/``)

Comment (ksaketou) [reply tests/functional/a/arguments_differ.py:30]: Yes I think typing is necessary if we want to check if the changes we did here actually work, that's why I added it here as well.

Comment (Pierre-Sassoulas) [reply tests/functional/a/arguments_differ.py:30]: Fair enough ! :)

**Comment by ksaketou** [comment]
My pleasure! So I have to wait for this PR to be merged in order to start working on the second pull request?

**Comment by Pierre-Sassoulas** [comment]
Well, you can disregard the error and start the second one, we will rebase on master once the first one is merged. I also see that there is a problem with mypy that you could check before that: https://results.pre-commit.ci/run/github/47671127/1620160302.btadc94JQO-yT6SZWp2aLg

**Comment by cdce8p** [comment]
> It look like we're getting a problem in the pipeline because of #4439. I'll merge as soon as someone fix it and the test pass when we rebase. Thank you for all the work on this !

The pylint problem here isn't related to the issue you linked. AFAIK it's because the annotation was assumed to have the `name` attribute. However that isn't always the case. It can also be of type `astroid.Attribute` which only has `attrname`.

https://github.com/PyCQA/pylint/blob/d05d2a18b57221c464c6d0799ba40bf66f306483/pylint/checkers/classes.py#L286-L290

**Comment by ksaketou** [comment]
I am working on the mypy errors, I think they might fix the problem. I had a small mistake on the parameter typing here.
https://github.com/PyCQA/pylint/blob/d05d2a18b57221c464c6d0799ba40bf66f306483/pylint/checkers/classes.py#L311-L312

**Pull request merged**
